Realm/RLMSyncSessionRefreshHandle.mm

```diff
@@ -53,20 +53,38 @@ - (instancetype)initWithFullURLPath:(NSString *)urlPath
 }
 
 - (void)invalidate {
-    [self.timer invalidate];
-    self.user = nil;
+    dispatch_async(dispatch_get_main_queue(), ^{
+        [self.timer invalidate];
+        self.user = nil;
+    });
 }
 
 - (void)scheduleRefreshTimer:(NSTimeInterval)fireTime {
-    static const NSInteger refreshBuffer = 10;
-    [self.timer invalidate];
-    self.timer = [[NSTimer alloc] initWithFireDate:[NSDate dateWithTimeIntervalSince1970:(fireTime - refreshBuffer)]
-                                          interval:0
-                                            target:self
-                                          selector:@selector(timerFired:)
-                                          userInfo:nil
-                                           repeats:NO];
-    [[NSRunLoop currentRunLoop] addTimer:self.timer forMode:NSDefaultRunLoopMode];
+    constexpr NSInteger REFRESH_BUFFER = 10;
+    // Schedule the timer on the main queue.
+    // It's very likely that this method will be run on a side thread, for example
+    // on the thread that runs `NSURLSession`'s completion blocks. We can't be
+    // guaranteed that there's an existing runloop on those threads, and we don't want
+    // to create and start a new one if one doesn't already exist.
```

Comment:
I'm inclined to leave it as-is for now, but I'll read up on dispatch sources and dispatch timers and we can refactor it if this becomes a problem.